tests/idem.rs

```diff
@@ -7,67 +7,91 @@
 // <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
-
+#![feature(std_misc)]
 extern crate rustfmt;
-
 use std::collections::HashMap;
 use std::fs;
 use std::io::Read;
-use std::sync::atomic;
 use rustfmt::*;
-
 // For now, the only supported regression tests are idempotent tests - the input and
 // output must match exactly.
 // FIXME(#28) would be good to check for error messages and fail on them, or at least report.
 #[test]
 fn idempotent_tests() {
     println!("Idempotent tests:");
-    FAILURES.store(0, atomic::Ordering::Relaxed);
-
     // Get all files in the tests/idem directory
     let files = fs::read_dir("tests/idem").unwrap();
+    let files2 = fs::read_dir("tests").unwrap();
+    let files3 = fs::read_dir("src/bin").unwrap();
     // For each file, run rustfmt and collect the output
     let mut count = 0;
-    for entry in files {
+    let mut fails = 0;
+    for entry in files.chain(files2).chain(files3) {
         let path = entry.unwrap().path();
         let file_name = path.to_str().unwrap();
+        if !file_name.ends_with(".rs") {
+            continue;
+        }
         println!("Testing '{}'...", file_name);
-        run(vec!["rustfmt".to_owned(), file_name.to_owned()], WriteMode::Return(HANDLE_RESULT));
+        match idempotent_check(file_name.to_owned()) {
+            Ok(()) => {},
+            Err(m) => {
+                print_mismatches(m);
+                fails += 1;
+            },
+        }
         count += 1;
     }
-    // And also dogfood ourselves!
+    // And also dogfood rustfmt!
     println!("Testing 'src/lib.rs'...");
-    run(vec!["rustfmt".to_string(), "src/lib.rs".to_string()],
-        WriteMode::Return(HANDLE_RESULT));
+    match idempotent_check("src/lib.rs".to_owned()) {
+        Ok(()) => {},
+        Err(m) => {
+            print_mismatches(m);
+            fails += 1;
+        },
+    }
     count += 1;
-
     // Display results
-    let fails = FAILURES.load(atomic::Ordering::Relaxed);
     println!("Ran {} idempotent tests; {} failures.", count, fails);
     assert!(fails == 0, "{} idempotent tests failed", fails);
 }
-
-// 'global' used by sys_tests and handle_result.
-static FAILURES: atomic::AtomicUsize = atomic::ATOMIC_USIZE_INIT;
+// Compare output to input.
+fn print_mismatches(result: HashMap<String, String>) {
+    for (file_name, fmt_text) in result {
+        println!("Mismatch in {}.", file_name);
+        println!("{}", fmt_text);
+    }
+}
 // Ick, just needed to get a &'static to handle_result.
 static HANDLE_RESULT: &'static Fn(HashMap<String, String>) = &handle_result;
-
+pub fn idempotent_check(filename: String) -> Result<(), HashMap<String, String>> {
+    use std::thread;
+    use std::fs;
+    use std::io::Read;
+    let args = vec!["rustfmt".to_owned(), filename];
+    thread::spawn(move || {
```

Comment:
I think you need a comment here to explain that you're using the thread here in order to get a result, rather than for concurrency. And explain how it works.